### DISCUSSION TITLE
iSmart App 2.0

### DISCUSSION THREAD

**Discussion by tisoft** [SAIC-iSmart-API/documentation#18] [General]
Hi,

The iSmart app in version 2.0 seems to use a new API endpoint: https://gateway-mg-eu.soimt.com/api.app/v1/oauth/token

Send to be a REST api. But the request and response body are encrypted. I haven't seen any ASN.1 messages there. 

Will have a look into it in the next few days. Maybe we can get rid of all the binary ASN.1 messages.

**Comment by tisoft** [comment]
Just a quick heads up: I managed to get the clear text from the encrypted bodies. Inside is a plain REST API with json.

Example response from unauthenticated endpoint `https://gateway-mg-eu.soimt.com/api.app/v1/user/country/list`:
```json
{"code":0,"data":[{"englishName":"Albania","localName":"Albania","countryCode":"AL","phoneCode":"355","description":"1","id":36},{"englishName":"Austria","localName":"Austria","countryCode":"AT","phoneCode":"43","description":"0","id":3},{"englishName":"Belgium","localName":"Belgium","countryCode":"BE","phoneCode":"32","description":"0","id":2},{"englishName":"Bosnia and Herzegovina","localName":"Bosnia and Herzegovina","countryCode":"BH","phoneCode":"387","description":"1","id":34},{"englishName":"Bulgaria","localName":"Bulgaria","countryCode":"BG","phoneCode":"359","description":"1","id":37},{"englishName":"Croatia","localName":"Croatia","countryCode":"CT","phoneCode":"385","description":"1","id":25},{"englishName":"Cyprus","localName":"Cyprus","countryCode":"CY","phoneCode":"357","description":"0","id":29},{"englishName":"Czech","localName":"Czech","countryCode":"CZ","phoneCode":"420","description":"1","id":22},{"englishName":"Denmark","localName":"Denmark","countryCode":"DK","phoneCode":"45","description":"0","id":1},{"englishName":"Estonia","localName":"Estonia","countryCode":"EST","phoneCode":"372","description":"1","id":38},{"englishName":"Finland","localName":"Finland","countryCode":"FI","phoneCode":"358","description":"0","id":18},{"englishName":"France","localName":"France","countryCode":"FR","phoneCode":"33","description":"0","id":11},{"englishName":"Germany","localName":"Germany","countryCode":"GE","phoneCode":"49","description":"0","id":12},{"englishName":"Greece","localName":"Greece","countryCode":"GR","phoneCode":"30","description":"1","id":19},{"englishName":"Hong Kong, China","localName":"Hong Kong，China","countryCode":"HK","phoneCode":"852","description":"0","id":43},{"englishName":"Hungary","localName":"Hungary","countryCode":"HU","phoneCode":"36","description":"1","id":21},{"englishName":"Iceland","localName":"Iceland","countryCode":"IS","phoneCode":"354","description":"0","id":17},{"englishName":"Ireland","localName":"Ireland","countryCode":"IE","phoneCode":"353","description":"0","id":16},{"englishName":"Israel","localName":"Israel","countryCode":"IL","phoneCode":"972","description":"0","id":26},{"englishName":"Italy","localName":"Italy","countryCode":"IT","phoneCode":"39","description":"0","id":10},{"englishName":"Latvia","localName":"Latvia","countryCode":"LV","phoneCode":"371","description":"1","id":39},{"englishName":"Lithuania","localName":"Lithuania","countryCode":"LTU","phoneCode":"370","description":"1","id":40},{"englishName":"Luxembourg","localName":"Luxembourg","countryCode":"LU","phoneCode":"352","description":"0","id":9},{"englishName":"Malaysia","localName":"Malaysia","countryCode":"MYS","phoneCode":"60","description":"3","id":50},{"englishName":"Malta","localName":"Malta","countryCode":"MT","phoneCode":"356","description":"0","id":28},{"englishName":"Moldova","localName":"Moldova","countryCode":"MD","phoneCode":"373","description":"1","id":41},{"englishName":"Montenegro","localName":"Montenegro","countryCode":"ME","phoneCode":"382","description":"1","id":35},{"englishName":"Nepal","localName":"Nepal","countryCode":"NPL","phoneCode":"977","description":"3","id":52},{"englishName":"Netherlands","localName":"Netherlands","countryCode":"NL","phoneCode":"31","description":"0","id":8},{"englishName":"North Macedonia","localName":"North Macedonia","countryCode":"NM","phoneCode":"389","description":"1","id":33},{"englishName":"Norway","localName":"Norway","countryCode":"NO","phoneCode":"47","description":"0","id":7},{"englishName":"Poland","localName":"Poland","countryCode":"PL","phoneCode":"48","description":"1","id":20},{"englishName":"Portugal","localName":"Portugal","countryCode":"PT","phoneCode":"351","description":"0","id":5},{"englishName":"Romania","localName":"Romania","countryCode":"ROM","phoneCode":"40","description":"1","id":42},{"englishName":"Serbia","localName":"Serbia","countryCode":"SRB","phoneCode":"381","description":"1","id":32},{"englishName":"Singapore","localName":"Singapore","countryCode":"SGP","phoneCode":"65","description":"3","id":51},{"englishName":"Slovakia","localName":"Slovakia","countryCode":"SK","phoneCode":"421","description":"0","id":23},{"englishName":"Slovenia","localName":"Slovenia","countryCode":"SI","phoneCode":"386","description":"0","id":24},{"englishName":"Spain","localName":"Spain","countryCode":"SP","phoneCode":"34","description":"0","id":13},{"englishName":"Sweden","localName":"Sweden","countryCode":"SE","phoneCode":"46","description":"0","id":4},{"englishName":"Switzerland","localName":"Switzerland","countryCode":"SZL","phoneCode":"41","description":"0","id":6},{"englishName":"UK","localName":"UK","countryCode":"UK","phoneCode":"44","description":"0","id":14},{"englishName":"Ukraine","localName":"Ukraine","countryCode":"UKR","phoneCode":"380","description":"1","id":44}],"message":"success"}
```

**Reply by tisoft** [reply]
The following endpoints seem to be available. Especially the sub account endpoints look interesting. Maybe that can be used to have multiple connections to a single car, so that a real co-existence of the official app and 3rd-party apps could be feasible.

## Endpoints:
- `GET` "fota/campaign"
- `POST` "fota/campaign/confirm"
- `GET` "fota/campaign/history"
- `GET` "fota/campaign/status"
- `POST` "vehicle/btkey/activation"
- `POST` "user/subaccount"
- `POST` "vehicle/btkey/authorization"
- `POST` "charging/batteryHeating"
- `GET` "charging/batteryHeating"
- `POST` "charging/charger/public/station/info"
- `POST` "charging/charger/public/station/list"
- `POST` "charging/setting/batteryHeating"
- `GET` "charging/setting/batteryHeating"
- `POST` "user/account/confirmLogin"
- `DELETE` "vehicle/btkey/authorization"
- `GET` "resource/oss/endpoint/get-file-attach"
- `POST` "user/feedback"
- `PUT` "user/account/forgotPassword"
- `GET` "vehicle/btkey/authorizedKeyList"
- `GET` "vehicle/statisticsBasicInfo"
- `GET` "user/subaccount/list"
- `GET` "vehicle/tbox/onlineStatus"
- `GET` "vehicle/btkey/userKeyList"
- `GET` "vehicle/apnAllowed"
- `GET` "vehicle/charging/mgmtData"
- `GET` "vehicle/statisticsInfo"
- `GET` "message/list"
- `PUT` "message/notificationCount"
- `GET` "message/notificationList"
- `PUT` "message/status"
- `GET` "message/unreadCount"
- `PUT` "user/subaccount"
- `GET` "navi/co2info"
- `PUT` "navi/co2info/itemCollect"
- `GET` "navi/co2info/ranking"
- `GET` "navi/co2info/supplementInfo"
- `PUT` "navi/co2info/supplementInfo/building/status"
- `PUT` "navi/co2info/supplementInfo/medal/status"
- `GET` "navi/dealer/list"
- `GET` "navi/driverbehavior/trip/list"
- `GET` "navi/driverbehavior/trip/route"
- `GET` "navi/outingplan"
- `POST` "navi/outingplan"
- `DELETE` "navi/outingplan"
- `GET` "navi/outingplan/list"
- `POST` "navi/poi/favorite"
- `DELETE` "navi/poi/favorite"
- `GET` "navi/poi/favorite/list"
- `PUT` "navi/poi/favorite"
- `POST` "navi/poi/send"
- `GET` "navi/roadAssistance/list"
- `POST` "oauth/token"
- `GET` "oauth/logout"
- `POST` "user/account/register"
- `POST` "user/account/scanLogin"
- `PUT` "user/subaccount/status"
- `POST` "vehicle/tbox/simCard"
- `POST` "user/subaccount/unbind"
- `POST` "user/account/unregister"
- `PUT` "vehicle/btkey/authorization"
- `PUT` "vehicle/btkey/authorization/status"
- `PUT` "user/account/userName"
- `PUT` "user/account/userInfo"
- `POST` "resource/oss/endpoint/upload-file-attach-by-name"
- `GET` "user/advertise/list"
- `POST` "user/anniversary"
- `DELETE` "user/anniversary"
- `GET` "user/anniversary/list"
- `PUT` "user/anniversary"
- `GET` "user/app/config"
- `GET` "user/app/skin/list"
- `GET` "user/app/skin/list" (with parameter "appSkinId")
- `GET` "user/app/upgradeInfo"
- `GET` "user/country/list"
- `POST` "user/gdprStatus"
- `GET` "user/handbook/list"
- `GET` "user/account/userInfo"
- `POST` "user/timezone"
- `GET` "user/timezone"
- `GET` "vehicle/alarm/list"
- `POST` "vehicle/alarm/release"
- `PUT` "vehicle/alarmSwitch"
- `GET` "vehicle/alarmSwitch"
- `POST` "vehicle/bind"
- `POST` "vehicle/bind/scanQrCode"
- `POST` "vehicle/bind/scanQrCode/nonInternet"
- `POST` "vehicle/bind/confirmBind"
- `GET` "vehicle/brandModel"
- `GET` "vehicle/ccInfo"
- `POST` "vehicle/charging/control"
- `GET` "vehicle/chargingPile/list"
- `POST` "vehicle/charging/ptcHeat"
- `POST` "vehicle/charging/reservation"
- `POST` "vehicle/charging/setting"
- `GET` "vehicle/charging/status"
- `POST` "vehicle/configSetting"
- `POST` "vehicle/control"
- `PUT` "vehicle/country"
- `GET` "vehicle/feature/list"
- `GET` "vehicle/geoFence"
- `POST` "vehicle/geoFence"
- `PUT` "vehicle/info"
- `GET` "vehicle/list"
- `GET` "vehicle/location"
- `GET` "vehicle/location" (with header "event-id")
- `POST` "vehicle/location/share"
- `DELETE` "vehicle/location/share"
- `GET` "vehicle/location/share/setting"
- `GET` "vehicle/location/share" (with query parameter "token")
- `PUT` "vehicle/pincode"
- `POST` "vehicle/pincode/verify"
- `GET` "vehicle/statisticsInfo"
- `POST` "vehicle/statisticsBasicInfo"
- `GET` "vehicle/status"
- `GET` "vehicle/status/classified"
- `PUT` "vehicle/unbind"
- `POST` "user/account/verificationCode"
- `GET` "user/account/verifyPassword"

**Reply by aserraric** [reply]
What is the encryption scheme? Symmetrical, I assume? Did you extract the key from the APK?

**Reply by tisoft** [reply]
It's symmetrical. The key is derived from the header parameters, so it's more an obfuscation than a real encryption.

I will document everything as soon as I have a working version, that can talk to the API.

**Comment by tisoft** [comment]
The API seems to be not yet fully implemented:

```
Unexpected response status: 500, body:{"code":500,"message":"\n### Error querying database.  Cause: java.sql.SQLSyntaxErrorException: Table 'iov_eu.tb_iot_device_info' doesn't exist\n### The error may exist in com/saicmotor/telematics/vehicle/mapper/TbIotDeviceInfoMapper.java (best guess)\n### The error may involve defaultParameterMap\n### The error occurred while setting parameters\n### SQL: SELECT id, vin, tbox_sn, last_activity_time, connect_status, create_by, create_date, last_update_by, last_update_date, row_version, is_valid, description, status FROM TB_IOT_DEVICE_INFO WHERE is_valid = 1 AND (vin = ?)\n### Cause: java.sql.SQLSyntaxErrorException: Table 'iov_eu.tb_iot_device_info' doesn't exist\n; bad SQL grammar []; nested exception is java.sql.SQLSyntaxErrorException: Table 'iov_eu.tb_iot_device_info' doesn't exist"}
```

**Comment by tisoft** [comment]
I can now talk to the new RESt API. Logging in and retrieving data from the car works.

Demo code:

https://github.com/SAIC-iSmart-API/saic-java-client/blob/api/saic-java-api-cli/src/main/java/net/heberling/ismart/cli/GetData2.java

The vehicle and charge data from the new API have the same structure as the corresponding endpoints from the old API. The ASN.1 structure is basically just converted to json.

There are several new API endpoints, some of them throw errors and are either not ready yet, or maybe disabled in the EU.

**Reply by BaxPlay360** [reply]
@tisoft So da ich es endlich hinbekommen habe, ist mir bei saic-mqtt-gateway noch ein problem aufgefallen: nach ca. 10 min überträgt er keine daten mehr und journalctl frezzed bzw. keine aktuellen infos mehr. ich muss es quasi restarten um wieder die daten abfragen zu können. kann an dsa irgendwie beheben oder alle 5 min einen auto restart einpflegen

**Reply by Tonno87** [reply]
Die Gateways fragen nach 10min keine Daten mehr ab um die Batterie zu schonen. Bei jedem Refresh wird das Radio kurz aktiviert um die Daten abzufragen. Das kann man natürlich nur solange machen bis die 12V Batterie leer ist.

Welche Daten möchtest du denn dauerhaft abfragen?

**Comment by kscholty** [comment]
This is really great work. Big thanks for the effort!
I used this as basis to develop a native EVCC vehicle connection. 
https://github.com/evcc-io/evcc/pull/12003

This works so far, but I noticed that the API calls often fail or deliver wrong data when the car is inactive. Especially in EVCC I see that the calls fail over and over, even though I notice the car waking up. This feels kind of dangerous, because it could drain the 12V battery.
Do you have an idea how to make the answers come in more reliably? Would it help to query quicker than every 30s as is minimum in EVCC?

**Reply by dougculnane** [reply]
Do you have an idea how to make the answers come in more reliably?

No. But the strategy implemented in openHAB is explained by it author here.
https://community.openhab.org/t/new-binding-integration-of-saic-mg-ismart-based-vehicles/140136/9?u=doug_culnane

Having a refresh that triggers the fast polling when charging starts keeps the battery level up to date without draining the battery.  However when charging stops the fast polling does too.

**Comment by dougculnane** [comment]
Thanks so much for this work. The openHAB binding has stopped working so I will try to use this as a basis to update it to new API. Will the code in the API branch get released as maven artifacts? This would be the easiest and cleanest way for me to reuse it.

Thanks again,
Doug

**Reply by tisoft** [reply]
Hi,
The code in the api branch is only a quick proof of concept. it is not usable as a real API currently. Unfortunately I don't have the capacity to bring it up to release quality at the moment.

But, if someone would open a PR with an updated implementation, I would happily merge it and make sure its available on Maven central, so that it can be used.

I f not, i do intend to come back to this and implement it myself, but can't give any time frame for it.

**Reply by dougculnane** [reply]
Thanks for the answer. I will try to find the time and have a play. I will try to get the openHAB binding working again and the code in PRs.

**Reply by tisoft** [reply]
Thanks!

I did update my GPG key, so builds work again and merged all open dependabot PRs, so the repo is ready for any new contributions. 😄
